File: fixtures/programming/scala/TaskService.scala
package com.example.tasks

import scala.collection.mutable

/** Status of a task. */
sealed trait TaskStatus
object TaskStatus {
  case object Pending extends TaskStatus
  case object Done extends TaskStatus
}

/**
 * A task with title and status.
 * @param id unique identifier
 * @param title short summary
 */
case class Task(id: Long, title: String, status: TaskStatus = TaskStatus.Pending) {
  /** Returns a formatted summary. */
  def summary: String = {
    val indicator = status match {
      case TaskStatus.Done => "✓"
      case _ => "○"
    }
    s"$indicator [$id] $title"
  }
}

/**
 * In-memory task service.
 *
 * The store is keyed by the id the service itself hands out, so a task
 * created elsewhere cannot collide with one created here.
 */
class TaskService {
  private val store = mutable.Map.empty[Long, Task]
  private var nextId = 1L

  /**
   * What one call to `complete` did.
   *
   * Nested in the service because the outcome only makes sense against
   * this service's store; a top-level `Outcome` would say nothing.
   */
  sealed trait Outcome {
    /** A one-word name for logs. */
    def label: String
  }

  /** The task was found and is now done. */
  case class Completed(task: Task) extends Outcome {
    def label: String = "completed"
  }

  /** Creates a new task. */
  def create(title: String): Task = {
    val task = Task(nextId, title)
    store(nextId) = task
    nextId += 1
    task
  }

  /** Marks a task as done. */
  def complete(id: Long): Option[Task] = {
    store.get(id).map { task =>
      val updated = task.copy(status = TaskStatus.Done)
      store(id) = updated
      updated
    }
  }

  /** Marks a task as done and reports what happened. */
  def completeReporting(id: Long): Option[Outcome] =
    complete(id).map(Completed.apply)

  /** Returns all pending tasks. */
  def pending: List[Task] =
    store.values.filter(_.status == TaskStatus.Pending).toList
}

File: fixtures/programming/ruby/task_service.rb
# frozen_string_literal: true

require_relative 'task'

# Service for managing tasks — wraps an in-memory store.
#
# The store is keyed by the id the service hands out itself, so a task
# built elsewhere can never collide with one created here.
class TaskService
  # Raised when an operation names a task that is not in the store.
  #
  # Nested inside the service because the id space it complains about is
  # the service's own; a top-level MissingTask would be ambiguous.
  class MissingTask < StandardError
    # @return [Integer] the id that was not found
    attr_reader :id

    # @param id [Integer] the id that was not found
    def initialize(id)
      @id = id
    end

    # A one-line description for a log.
    # @return [String]
    def message
      "no task with id #{@id}"
    end
  end

  def initialize
    @tasks = {}
    @next_id = 1
  end

  # Adds a task with the given title and returns it.
  # @param title [String] the task title
  # @return [Task]
  def add(title)
    task = Task.new(id: @next_id, title: title)
    @tasks[@next_id] = task
    @next_id += 1
    task
  end

  # Marks a task as done by ID.
  # @param id [Integer]
  # @return [Boolean] true if found
  def complete(id)
    task = @tasks[id]
    return false unless task

    task.complete!
    true
  end

  # Marks a task as done, raising when it is not there.
  # @param id [Integer]
  # @raise [MissingTask] when no task has that id
  # @return [Task]
  def complete!(id)
    raise MissingTask.new(id) unless @tasks.key?(id)

    @tasks[id].tap(&:complete!)
  end

  # Returns all pending tasks.
  # @return [Array<Task>]
  def pending
    @tasks.values.reject(&:done?)
  end

  # Yields each pending task, or returns an enumerator without a block.
  # @return [Enumerator, nil]
  def each_pending
    return to_enum(:each_pending) unless block_given?

    pending.each { |task| yield task }
  end

  # The title of the most recently added task, if there is one.
  # @return [String, nil]
  def latest_title
    @tasks.values.last&.title
  end

  # Renders the backlog as a block of text.
  # @return [String]
  def report
    <<~REPORT
      #{self.class.name} — #{pending.length} pending
      #{pending.map { |t| "- #{t.title}" }.join("\n")}
    REPORT
  end

  alias to_s report

  # Builds a service preloaded with the given titles.
  # @param titles [Array<String>]
  # @return [TaskService]
  def self.seeded(titles)
    titles.each_with_object(new) { |title, service| service.add(title) }
  end
end
